File: frontend/lib/models/borrowing.dart
import 'dart:convert';
import 'package:logger/logger.dart';

final logger = Logger();

class Borrowing {
  final int id;
  final int borrowerId;
  final String borrowerName;
  final String borrowerEmail;
  final int? technicianId;
  final String? technicianName;
  final DateTime? technicianApprovedAt;
  final int? adminId;
  final String? adminName;
  final DateTime? adminApprovedAt;
  final List<dynamic> chemicals;
  final List<dynamic> equipment;
  final String purpose;
  final String researchDetails;
  final DateTime borrowDate;
  final DateTime returnDate;
  final DateTime visitDate;
  final String visitTime;
  final String status;
  final String? notes;
  final String? rejectionReason;
  final DateTime createdAt;
  final DateTime updatedAt;

  // ✅ NEW: Return confirmation fields
  final bool returned;
  final DateTime? actualReturnDate;
  final Map<String, dynamic>? equipmentCondition;
  final String? returnNotes;
  final int? returnConfirmedBy;

  // Add student information fields
  final String? university;
  final String? educationLevel;
  final String? registrationNumber;
  final String? studentNumber;
  final int? currentYear;
  final String? semester;
  final String? borrowerEmailContact;
  final String? borrowerContact;

  Borrowing({
    required this.id,
    required this.borrowerId,
    required this.borrowerName,
    required this.borrowerEmail,
    this.technicianId,
    this.technicianName,
    this.technicianApprovedAt,
    this.adminId,
    this.adminName,
    this.adminApprovedAt,
    required this.chemicals,
    required this.equipment,
    required this.purpose,
    required this.researchDetails,
    required this.borrowDate,
    required this.returnDate,
    required this.visitDate,
    required this.visitTime,
    required this.status,
    this.notes,
    this.rejectionReason,
    required this.createdAt,
    required this.updatedAt,
    // ✅ NEW: Return confirmation fields
    this.returned = false,
    this.actualReturnDate,
    this.equipmentCondition,
    this.returnNotes,
    this.returnConfirmedBy,
    // Add student information fields
    this.university,
    this.educationLevel,
    this.registrationNumber,
    this.studentNumber,
    this.currentYear,
    this.semester,
    this.borrowerEmailContact,
    this.borrowerContact,
  });

  factory Borrowing.fromJson(Map<String, dynamic> json) {
    // ✅ Handle equipment condition parsing
    Map<String, dynamic>? equipmentCondition;
    if (json['equipment_condition'] != null) {
      try {
        if (json['equipment_condition'] is String) {
          equipmentCondition = jsonDecode(json['equipment_condition']);
        } else if (json['equipment_condition'] is Map) {
          equipmentCondition = json['equipment_condition'];
        }
      } catch (e) {
        logger.d('Error parsing equipment condition: $e');
      }
    }

    return Borrowing(
      id: json['id'],
      borrowerId: json['borrower_id'],
      borrowerName: json['borrower_name'] ?? '',
      borrowerEmail: json['borrower_email'] ?? '',
      technicianId: json['technician_id'],
      technicianName: json['technician_name'],
      technicianApprovedAt: json['technician_approved_at'] != null
          ? DateTime.parse(json['technician_approved_at'])
          : null,
      adminId: json['admin_id'],
      adminName: json['admin_name'],
      adminApprovedAt: json['admin_approved_at'] != null
          ? DateTime.parse(json['admin_approved_at'])
          : null,
      chemicals: json['chemicals'] ?? [],
      equipment: json['equipment'] ?? [],
      purpose: json['purpose'],
      researchDetails: json['research_details'],
      borrowDate: DateTime.parse(json['borrow_date']),
      returnDate: DateTime.parse(json['return_date']),
      visitDate: DateTime.parse(json['visit_date']),
      visitTime: json['visit_time'],
      status: json['status'],
      notes: json['notes'],
      rejectionReason: json['rejection_reason'],
      createdAt: DateTime.parse(json['created_at']),
      updatedAt: DateTime.parse(json['updated_at'] ?? json['created_at']),
      // ✅ NEW: Return confirmation fields
      returned: json['returned'] ?? false,
      actualReturnDate: json['actual_return_date'] != null
          ? DateTime.parse(json['actual_return_date'])
          : null,
      equipmentCondition: equipmentCondition,
      returnNotes: json['return_notes'],
      returnConfirmedBy: json['return_confirmed_by'],
      // Add student information fields
      university: json['university'],
      educationLevel: json['education_level'],
      registrationNumber: json['registration_number'],
      studentNumber: json['student_number'],
      currentYear: json['current_year'],
      semester: json['semester'],
      borrowerEmailContact: json['borrower_email'],
      borrowerContact: json['borrower_contact'],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'borrower_id': borrowerId,
      'borrower_name': borrowerName,
      'borrower_email': borrowerEmail,
      'technician_id': technicianId,
      'technician_name': technicianName,
      'technician_approved_at': technicianApprovedAt?.toIso8601String(),
      'admin_id': adminId,
      'admin_name': adminName,
      'admin_approved_at': adminApprovedAt?.toIso8601String(),
      'chemicals': chemicals,
      'equipment': equipment,
      'purpose': purpose,
      'research_details': researchDetails,
      'borrow_date': borrowDate.toIso8601String(),
      'return_date': returnDate.toIso8601String(),
      'visit_date': visitDate.toIso8601String(),
      'visit_time': visitTime,
      'status': status,
      'notes': notes,
      'rejection_reason': rejectionReason,
      'created_at': createdAt.toIso8601String(),
      'updated_at': updatedAt.toIso8601String(),
      // ✅ NEW: Return confirmation fields
      'returned': returned,
      'actual_return_date': actualReturnDate?.toIso8601String(),
      'equipment_condition':
          equipmentCondition != null ? jsonEncode(equipmentCondition) : null,
      'return_notes': returnNotes,
      'return_confirmed_by': returnConfirmedBy,
      // Add student information fields
      'university': university,
      'education_level': educationLevel,
      'registration_number': registrationNumber,
      'student_number': studentNumber,
      'current_year': currentYear,
      'semester': semester,
      // ignore: equal_keys_in_map
      'borrower_email': borrowerEmailContact,
      'borrower_contact': borrowerContact,
    };
  }
}

File: backend/models/Borrowing.js
const db = require('../config/db');

class Borrowing {
  static async create(borrowingData) {
    const { 
      borrower_id, 
      chemicals, 
      equipment, 
      purpose, 
      research_details, 
      borrow_date, 
      return_date, 
      visit_date, 
      visit_time,
      // Add student information fields
      university,
      education_level,
      registration_number,
      student_number,
      current_year,
      semester,
      borrower_email,
      borrower_contact
    } = borrowingData;
    
    // Ensure chemicals and equipment are properly formatted as JSON
    const formattedChemicals = Array.isArray(chemicals) ? JSON.stringify(chemicals) : '[]';
    const formattedEquipment = Array.isArray(equipment) ? JSON.stringify(equipment) : '[]';
    
    console.log('Creating borrowing with data:', {
      borrower_id,
      chemicals: formattedChemicals,
      equipment: formattedEquipment,
      purpose,
      research_details,
      borrow_date,
      return_date,
      visit_date,
      visit_time,
      university,
      education_level,
      registration_number,
      student_number,
      current_year,
      semester,
      borrower_email,
      borrower_contact
    });
    
    const result = await db.query(
      `INSERT INTO borrowings (
         borrower_id, chemicals, equipment, purpose, research_details, 
         borrow_date, return_date, visit_date, visit_time, status,
         university, education_level, registration_number, student_number,
         current_year, semester, borrower_email, borrower_contact
       )
       VALUES (
         $1, $2, $3, $4, $5, $6, $7, $8, $9, 'pending',
         $10, $11, $12, $13, $14, $15, $16, $17
       ) RETURNING *`,
      [
        borrower_id, formattedChemicals, formattedEquipment, purpose, research_details,
        borrow_date, return_date, visit_date, visit_time,
        university, education_level, registration_number, student_number,
        current_year, semester, borrower_email, borrower_contact
      ]
    );
    
    return result.rows[0];
  }

  static async findAll(filters = {}) {
    let query = `
      SELECT b.*, 
             u.name as borrower_name, 
             u.email as borrower_email,
             t.name as technician_name,
             a.name as admin_name
      FROM borrowings b 
      JOIN users u ON b.borrower_id = u.id 
      LEFT JOIN users t ON b.technician_id = t.id
      LEFT JOIN users a ON b.admin_id = a.id
      WHERE 1=1
    `;
    const params = [];
    let paramIndex = 1;

    if (filters.status) {
      query += ` AND b.status = $${paramIndex}`;
      params.push(filters.status);
      paramIndex++;
    }

    if (filters.borrower_id) {
      query += ` AND b.borrower_id = $${paramIndex}`;
      params.push(filters.borrower_id);
      paramIndex++;
    }

    query += ' ORDER BY b.created_at DESC';

    try {
      const result = await db.query(query, params);
      
      // Parse JSON fields
      const borrowings = result.rows.map(row => ({
        ...row,
        chemicals: typeof row.chemicals === 'string' ? JSON.parse(row.chemicals) : row.chemicals || [],
        equipment: typeof row.equipment === 'string' ? JSON.parse(row.equipment) : row.equipment || [],
        // Include student information fields
        university: row.university,
        education_level: row.education_level,
        registration_number: row.registration_number,
        student_number: row.student_number,
        current_year: row.current_year,
        semester: row.semester,
        borrower_email: row.borrower_email,
        borrower_contact: row.borrower_contact,
        // ✅ Parse equipment_condition if it exists
        equipment_condition: row.equipment_condition ? 
          (typeof row.equipment_condition === 'string' ? JSON.parse(row.equipment_condition) : row.equipment_condition) : null
      }));
      
      return borrowings;
    } catch (error) {
      console.error('Error in Borrowing.findAll:', error);
      throw error;
    }
  }

  static async findById(id) {
    try {
      const result = await db.query(
        `SELECT b.*, 
                u.name as borrower_name, 
                u.email as borrower_email,
                t.name as technician_name,
                a.name as admin_name
         FROM borrowings b 
         JOIN users u ON b.borrower_id = u.id 
         LEFT JOIN users t ON b.technician_id = t.id
         LEFT JOIN users a ON b.admin_id = a.id
         WHERE b.id = $1`,
        [id]
      );
      
      if (result.rows.length === 0) {
        return null;
      }
      
      const row = result.rows[0];
      return {
        ...row,
        chemicals: typeof row.chemicals === 'string' ? JSON.parse(row.chemicals) : row.chemicals || [],
        equipment: typeof row.equipment === 'string' ? JSON.parse(row.equipment) : row.equipment || [],
        // Include student information fields
        university: row.university,
        education_level: row.education_level,
        registration_number: row.registration_number,
        student_number: row.student_number,
        current_year: row.current_year,
        semester: row.semester,
        borrower_email: row.borrower_email,
        borrower_contact: row.borrower_contact,
        // ✅ Parse equipment_condition if it exists
        equipment_condition: row.equipment_condition ? 
          (typeof row.equipment_condition === 'string' ? JSON.parse(row.equipment_condition) : row.equipment_condition) : null
      };
    } catch (error) {
      console.error('Error in Borrowing.findById:', error);
      throw error;
    }
  }

  static async updateStatus(id, status, updaterId, updaterRole, notes = null, rejectionReason = null) {
    let query;
    let params;
    
    console.log('Updating borrowing status:', { id, status, updaterId, updaterRole, notes, rejectionReason });
    
    try {
      if (status === 'approved' && updaterRole === 'technician') {
        // Technician approval - set technician info
        query = `UPDATE borrowings 
                 SET status = $1, 
                     technician_id = $2, 
                     technician_approved_at = NOW(),
                     notes = $3,
                     updated_at = NOW() 
                 WHERE id = $4 RETURNING *`;
        params = [status, updaterId, notes, id];
      } else if (status === 'approved' && updaterRole === 'admin') {
        // Admin approval - set admin info (final approval)
        query = `UPDATE borrowings 
                 SET status = $1, 
                     admin_id = $2, 
                     admin_approved_at = NOW(),
                     notes = $3,
                     updated_at = NOW() 
                 WHERE id = $4 RETURNING *`;
        params = [status, updaterId, notes, id];
      } else if (status === 'rejected') {
        // Rejection by either technician or admin
        const roleField = updaterRole === 'technician' ? 'technician_id' : 'admin_id';
        const timeField = updaterRole === 'technician' ? 'technician_approved_at' : 'admin_approved_at';
        
        query = `UPDATE borrowings 
                 SET status = $1, 
                     ${roleField} = $2, 
                     ${timeField} = NOW(),
                     rejection_reason = $3,
                     notes = $4,
                     updated_at = NOW() 
                 WHERE id = $5 RETURNING *`;
        params = [status, updaterId, rejectionReason, notes, id];
      } else {
        // Other status updates (returned, etc.)
        query = `UPDATE borrowings 
                 SET status = $1, 
                     notes = $2,
                     updated_at = NOW() 
                 WHERE id = $3 RETURNING *`;
        params = [status, notes, id];
      }

      const result = await db.query(query, params);
      
      if (result.rows.length === 0) {
        return null;
      }
      
      const row = result.rows[0];
      return {
        ...row,
        chemicals: typeof row.chemicals === 'string' ? JSON.parse(row.chemicals) : row.chemicals || [],
        equipment: typeof row.equipment === 'string' ? JSON.parse(row.equipment) : row.equipment || [],
        // Include student information fields
        university: row.university,
        education_level: row.education_level,
        registration_number: row.registration_number,
        student_number: row.student_number,
        current_year: row.current_year,
        semester: row.semester,
        borrower_email: row.borrower_email,
        borrower_contact: row.borrower_contact,
        // ✅ Parse equipment_condition if it exists
        equipment_condition: row.equipment_condition ? 
          (typeof row.equipment_condition === 'string' ? JSON.parse(row.equipment_condition) : row.equipment_condition) : null
      };
    } catch (error) {
      console.error('Error in Borrowing.updateStatus:', error);
      throw error;
    }
  }

  static async getPendingRequests() {
    try {
      const result = await db.query(
        `SELECT b.*, u.name as borrower_name, u.email as borrower_email 
         FROM borrowings b 
         JOIN users u ON b.borrower_id = u.id 
         WHERE b.status = 'pending'
         ORDER BY b.created_at ASC`
      );
      
      const pendingRequests = result.rows.map(row => ({
        ...row,
        chemicals: typeof row.chemicals === 'string' ? JSON.parse(row.chemicals) : row.chemicals || [],
        equipment: typeof row.equipment === 'string' ? JSON.parse(row.equipment) : row.equipment || [],
        // Include student information fields
        university: row.university,
        education_level: row.education_level,
        registration_number: row.registration_number,
        student_number: row.student_number,
        current_year: row.current_year,
        semester: row.semester,
        borrower_email: row.borrower_email,
        borrower_contact: row.borrower_contact,
        // ✅ Parse equipment_condition if it exists
        equipment_condition: row.equipment_condition ? 
          (typeof row.equipment_condition === 'string' ? JSON.parse(row.equipment_condition) : row.equipment_condition) : null
      }));
      
      return pendingRequests;
    } catch (error) {
      console.error('Error in Borrowing.getPendingRequests:', error);
      // Return empty array instead of throwing error
      return [];
    }
  }

  static async getPendingRequestsCount() {
    try {
      const result = await db.query(
        `SELECT COUNT(*) as count
         FROM borrowings 
         WHERE status = 'pending'`
      );
      
      return parseInt(result.rows[0].count);
    } catch (error) {
      console.error('Error in Borrowing.getPendingRequestsCount:', error);
      // Return 0 instead of throwing error
      return 0;
    }
  }

  // ✅ FIXED: markAsReturned method with proper validation and error handling
  static async markAsReturned(borrowingId, returnData, confirmedByUserId) {
    const { equipmentCondition, returnNotes } = returnData;
    
    console.log('=== MARK AS RETURNED METHOD CALLED ===');
    console.log('Borrowing ID:', borrowingId);
    console.log('Equipment Condition:', equipmentCondition);
    console.log('Return Notes:', returnNotes);
    console.log('Confirmed By User ID:', confirmedByUserId);
    
    // ✅ VALIDATE: Check if required data exists
    if (!borrowingId) {
      throw new Error('Borrowing ID is required');
    }
    
    if (!equipmentCondition) {
      throw new Error('Equipment condition data is required');
    }
    
    if (!confirmedByUserId) {
      throw new Error('Confirmed by user ID is required');
    }

    // ✅ VALIDATE: Check if equipmentCondition is an object
    if (typeof equipmentCondition !== 'object' || equipmentCondition === null) {
      throw new Error('Equipment condition must be a valid object');
    }

    try {
      const result = await db.query(
        `UPDATE borrowings 
         SET returned = true,
             actual_return_date = NOW(),
             equipment_condition = $1,
             return_notes = $2,
             return_confirmed_by = $3,
             status = 'returned',
             updated_at = NOW()
         WHERE id = $4 
         RETURNING *`,
        [
          JSON.stringify(equipmentCondition), 
          returnNotes || '', 
          confirmedByUserId, 
          borrowingId
        ]
      );
      
      console.log('Database update result rows:', result.rows.length);

      if (result.rows.length === 0) {
        throw new Error('Borrowing not found');
      }
      
      const row = result.rows[0];
      console.log('Successfully updated borrowing:', row);
      
      return {
        ...row,
        chemicals: typeof row.chemicals === 'string' ? JSON.parse(row.chemicals) : row.chemicals || [],
        equipment: typeof row.equipment === 'string' ? JSON.parse(row.equipment) : row.equipment || [],
        equipment_condition: row.equipment_condition ? 
          (typeof row.equipment_condition === 'string' ? JSON.parse(row.equipment_condition) : row.equipment_condition) : null
      };
    } catch (error) {
      console.error('Database error in markAsReturned:', error);
      throw new Error(`Database operation failed: ${error.message}`);
    }
  }

  // ✅ FIXED: getActiveBorrowings method
  static async getActiveBorrowings() {
    try {
      const result = await db.query(
        `SELECT b.*, 
                u.name as borrower_name, 
                u.email as borrower_email,
                tech.name as technician_name
         FROM borrowings b
         JOIN users u ON b.borrower_id = u.id
         LEFT JOIN users tech ON b.technician_id = tech.id
         WHERE b.status = 'approved' 
         AND b.returned = false
         ORDER BY b.created_at DESC`
      );
      
      return result.rows.map(row => ({
        ...row,
        chemicals: typeof row.chemicals === 'string' ? JSON.parse(row.chemicals) : row.chemicals || [],
        equipment: typeof row.equipment === 'string' ? JSON.parse(row.equipment) : row.equipment || [],
        equipment_condition: row.equipment_condition ? 
          (typeof row.equipment_condition === 'string' ? JSON.parse(row.equipment_condition) : row.equipment_condition) : null
      }));
    } catch (error) {
      console.error('Error fetching active borrowings:', error);
      throw error;
    }
  }

  static async getOverdue() {
    try {
      // Check if required columns exist
      const columnCheck = await db.query(`
        SELECT column_name 
        FROM information_schema.columns 
        WHERE table_name = 'borrowings' 
        AND column_name IN ('technician_id', 'admin_id', 'rejection_reason', 'university', 'education_level', 'registration_number', 'student_number', 'current_year', 'semester', 'borrower_email', 'borrower_contact')
      `);
      
      const existingColumns = columnCheck.rows.map(row => row.column_name);
      const hasTechId = existingColumns.includes('technician_id');
      const hasAdminId = existingColumns.includes('admin_id');
      const hasRejectionReason = existingColumns.includes('rejection_reason');
      
      let selectClause = 'b.*';
      let joinClause = '';
      
      if (hasTechId) {
        selectClause += ', t.name as technician_name';
        joinClause += ' LEFT JOIN users t ON b.technician_id = t.id';
      }
      
      if (hasAdminId) {
        selectClause += ', a.name as admin_name';
        joinClause += ' LEFT JOIN users a ON b.admin_id = a.id';
      }
      
      const query = `
        SELECT ${selectClause}, u.name as borrower_name, u.email as borrower_email 
        FROM borrowings b 
        JOIN users u ON b.borrower_id = u.id 
        ${joinClause}
        WHERE b.status = 'approved' AND b.return_date < CURRENT_DATE
      `;
      
      const result = await db.query(query);
      
      // Parse JSON fields
      const overdueBorrowings = result.rows.map(row => ({
        ...row,
        chemicals: typeof row.chemicals === 'string' ? JSON.parse(row.chemicals) : row.chemicals || [],
        equipment: typeof row.equipment === 'string' ? JSON.parse(row.equipment) : row.equipment || [],
        // Include student information fields
        university: row.university,
        education_level: row.education_level,
        registration_number: row.registration_number,
        student_number: row.student_number,
        current_year: row.current_year,
        semester: row.semester,
        borrower_email: row.borrower_email,
        borrower_contact: row.borrower_contact,
        // ✅ Parse equipment_condition if it exists
        equipment_condition: row.equipment_condition ? 
          (typeof row.equipment_condition === 'string' ? JSON.parse(row.equipment_condition) : row.equipment_condition) : null
      }));
      
      return overdueBorrowings;
    } catch (error) {
      console.error('Error in Borrowing.getOverdue:', error);
      // Return empty array instead of throwing error
      return [];
    }
  }
}

module.exports = Borrowing;